Add support for cross-site communication using Openshift Routes

pruivo opened the pull request:
Enable Openshift Route for cross-site communication. 

Currently, I'm unable to add tests for CI. :/

ryanemerson commented on controllers/xsite_config.go:
```diff
@@ -112,18 +139,40 @@ func appendRemoteLocation(ctx context.Context, infinispan *ispnv1.Infinispan, re
+	} else {
+		// Try searching for the Route
+		siteRoute := &routev1.Route{}
+		err = remoteKubernetes.Client.Get(context.TODO(), types.NamespacedName{Name: remoteRouteName, Namespace: remoteNamespace}, siteRoute)
```
We need to use the `ctx` provided as a parameter.

pruivo replied:
fixed!

ryanemerson commented on Makefile:
```diff
@@ -134,7 +134,7 @@ generate: controller-gen rice
 ## Build the operator image
 operator-build: manager
-	$(CONTAINER_TOOL) build -t $(IMG) .
+	$(CONTAINER_TOOL) build --build-arg VERSION=$(VERSION) -t $(IMG) .
```
Why is this necessary?

pruivo replied:
There is a `VERSION` argument in the Dockerfile that is not set and it does not have a default value. 
Isn't an issue?

Podman does not like it
```
$ make operator-build CONTAINER_TOOL=podman
/home/pedro/Work/infinispan-operator/bin/controller-gen object paths="./..."
/home/pedro/Work/infinispan-operator/bin/rice embed-go -i controllers/dependencies.go -i controllers/grafana.go
find . -type f -name 'rice-box.go' -exec sed -i "s|time.Unix(.*, 0)|time.Unix(1620137619, 0)|" {} \;
go fmt ./...
go vet ./...
go build -o bin/manager main.go
podman build -t quay.io/infinispan/operator:latest .
WARN[0001] missing "VERSION" build argument. Try adding "--build-arg VERSION=<VALUE>" to the command line 
```

pruivo replied:
actually, `VERSION` is set in the base image.
Adding `RUN echo ${VERSION}` I got (with docker)
```
Step 1/11 : FROM registry.access.redhat.com/ubi8/go-toolset:1.15.14 AS build
...
Step 3/11 : RUN echo ${VERSION}
 ---> Running in 6aeb06d29666
1.15.14
```
Is it intentional?

ryanemerson replied:
> There is a `VERSION` argument in the Dockerfile that is not set and it does not have a default value.
> Isn't an issue?

No it is, good spot :+1: 

> actually, `VERSION` is set in the base image. Adding `RUN echo ${VERSION}` I got (with docker)
> Is it intentional?

Nope :sweat_smile:

ryanemerson commented:
> Currently, I'm unable to add tests for CI. :/

You could add a test that works with Openshift. We can then perform a quick check at the start of the test method to determine if the k8s flavour supports Routes and if not calls `t.Skip()`.

pruivo commented:
> We can then perform a quick check at the start of the test method to determine if the k8s flavour supports Routes

I need time to investigate how to do it :)

ryanemerson commented:
> I need time to investigate how to do it :)

We already have a method to determine if a GVK is supported on the k8s platform: https://github.com/infinispan/infinispan-operator/blob/main/pkg/kubernetes/kubernetes.go#L72-L92

ryanemerson commented:
@pruivo I knew I had already implemented something similar for the testsuite, but couldn't find it ... you can use the `tutils.SkipForExposeType` to ensure a test is only executed when the `EXPOSE_SERVICE_TYPE=Route`.

https://github.com/ryanemerson/infinispan-operator/commit/c8219e311f827d226a5a34c53427441e398f3833

pruivo commented:
it looks like that commit is not available in the operator. Do you mind if I cherry-pick it to this PR?

ryanemerson commented:
@pruivo That explains why I couldn't find the code :upside_down_face:.  Go for it!

Crumby commented on config/crd/bases/infinispan.org_infinispans.yaml:
```diff
@@ -921,13 +921,18 @@ spec:
                               port:
                                 format: int32
                                 type: integer
+                              routerHostName:
```
Here is inconsistency between field name and the description. While the field is `routerHostName` the description mentions `RouteHostName` (route vs router), probably should be `route`

Crumby commented on controllers/infinispan_service_controller.go:
```diff
@@ -538,6 +554,39 @@ func computeSiteService(ispn *ispnv1.Infinispan) *corev1.Service {
+			Labels:    LabelsResource(ispn.Name, "infinispan-route-xsite"),
+		},
+		Spec: routev1.RouteSpec{
+			Host: host,
```
What if the `host` is empty?

Crumby replied:
Right, I was more wondering whether OpenShift won't complain if you send a request with explicit `host: nil`, compared to "not mentioning" `host` at all. But if it works then good, it's definitely nicer then having if block.

Crumby commented on test/e2e/xsite/xsite_test.go:
```diff
@@ -231,6 +231,16 @@ func TestDefaultTLSLoadBalancerWithPort(t *testing.T) {
 }
 
+func TestCrossSiteViewOpenshiftRoute(t *testing.T) {
+	tutils.SkipForExposeType(t, ispnv1.ExposeTypeNodePort, ispnv1.ExposeTypeLoadBalancer)
```
Maybe skipping based on whether `Route` API is available would be more consistent with the rest of the test rather then on test suite configuration? (We are not skipping LoadBalancer tests just because NodePort expose type is configured)

pruivo replied:
any pointers how to do that?

ryanemerson replied:
We can make use of the `IsGroupVersionSupported` method like [here](https://github.com/infinispan/infinispan-operator/blob/main/test/e2e/utils/common.go#L216) to see if it's supported and if not call skip on the test.

Crumby commented on test/e2e/xsite/xsite_test.go:
```diff
@@ -231,6 +231,16 @@ func TestDefaultTLSLoadBalancerWithPort(t *testing.T) {
 
+func TestCrossSiteViewOpenshiftRoute(t *testing.T) {
+	tutils.SkipForExposeType(t, ispnv1.ExposeTypeNodePort, ispnv1.ExposeTypeLoadBalancer)
+	testCrossSiteView(t, true, ispnv1.CrossSiteSchemeTypeOpenShift, ispnv1.CrossSiteExposeTypeRoute, 0, 1, NoTLS, nil)
```
`NoTLS` option is confusing. Cross site over Route won't work without TLS

pruivo replied:
I'm going to remove this test. It does not make sense to me :rofl:

Crumby commented on test/e2e/xsite/xsite_test.go:
```diff
@@ -231,6 +231,16 @@ func TestDefaultTLSLoadBalancerWithPort(t *testing.T) {
 	testCrossSiteView(t, true, ispnv1.CrossSiteSchemeTypeOpenShift, ispnv1.CrossSiteExposeTypeLoadBalancer, 1443, 1, DefaultTLS, nil)
 }
 
+func TestCrossSiteViewOpenshiftRoute(t *testing.T) {
```
Having internal version of the test for smoke testing on a single cluster would be nice

pruivo replied:
that is useless. Cross-site in a single cluster uses the cluster IP directly and not the Routes (same for NodePort & LoadBalancer)

Crumby replied:
Right, I incorrectly though we have "single cluster" version of the tests for NodePort and LoadBalancer too.

pruivo commented:
Updated!
* Fixed `TestCrossSiteViewKubernetesLoadBalancer`
* Tested the skip() method
```
=== RUN   TestDefaultTLSOpenshiftRoute
    xsite_test.go:297: Route not available. Skipping test
--- SKIP: TestDefaultTLSOpenshiftRoute (6.09s)
```

pruivo commented:
green :partying_face:

ryanemerson commented:
Thanks @pruivo
